### ISSUE TITLE
Feature Request: Allow namespace rather than kubelet to supply the added DNS client config

### ISSUE THREAD

**Issue by ardnaxelarak** [kubernetes/kubernetes#31017]
The use case for this new feature is a Kubernetes operator who is
providing a multi-tenant service based on namespaces.  Each namespace
belongs to one tenant.  One tenant should not have a view of the whole
cluster; rather the tenant's view should be constricted properly.

There are a few facets to this problem.  One of them is the DNS client
configuration applied in the tenant's pods.  The k8s API has an optional
`dnsPolicy` field in a `PodSpec`, and that field's value can be either
`ClusterFirst` or `Default`; the default behavior is that of
`ClusterFirst`.  The `Default` setting means that the pod uses its
host's DNS client configuration.  The `ClusterFirst` setting means
that the pod's DNS client configuration is constructed from the host's
by conditionally prepending some stuff.  The kubelet's `--cluster-dns`
parameter, if supplied, is prepended to the DNS server list.  The
domain search list is prepended by the kubelet's `--cluster-domain`
parameter if there is one, and then `svc.cluster.local`
`cluster.local` (where the `cluster.local` part can be varied by
another kubelet parameter) unconditionally.

The idea in this proposal is to allow a namespace to override the
behavior of the `ClusterFirst` setting to make it mean that the stuff
prepended comes from a new namespace field rather than kubelet
parameters.  We introduce into `NamespaceSpec` an optional field
`dnsClientConfig` of a new type `DNSClientConfig`.  A
`DNSClientConfig` holds a list of DNS server addresses and a list of
strings (search domains).  If the `dnsClientConfig` is present in a
namespace then for every pod in that namespace the `ClusterFirst`
behavior is to derive the pod's DNS client config by prepending the
contents of the namespace's `DNSClientConfig` (rather than the
settings derived from the kubelet parameters) to the config from the
host.

**Comment by MikeSpreitzer** [comment]
Sorry, I botched the description of the current behavior a little.  Or maybe it varies from release to release.  What I found in Kubernetes 1.3.5 is that the `ClusterFirst` behavior ignores the host's DNS client configuration and gives the pod only settings based on the kubelet arguments.  And regarding `--cluster-domain` the behavior is to set the pod's search list to `["default.svc.fubar", "svc.fubar", "fubar"]` when `--cluster-domain=fubar` and to `[]` when the kubelet is given no `--cluster-domain`.

The point of this feature proposal is that if the namespace has a `dnsClientConfig` then it is used in place of the stuff derived from kubelet parameters, to provide an appropriately scoped view of the cluster.
